[CGData] Global Merge Functions

kyulee-com opened the pull request:
This implements a global function merging pass. Unlike traditional function merging passes that use IR comparators, this pass employs a structurally stable hash to identify similar functions while ignoring certain constant operands. These ignored constants are tracked and encoded into a stable function summary. When merging, instead of explicitly folding similar functions and their call sites, we form a merging instance by supplying different parameters via thunks. The actual size reduction occurs when identically created merging instances are folded by the linker.

Currently, this pass is wired to llvm-lto2, enabled by the `-enable-global-merge-func` flag.
In a local merging mode, the analysis and merging steps occur sequentially within a module:
 - `analyze`: Collects stable function hashes and tracks locations of ignored constant operands.
 - `finalize`: Identifies merge candidates with matching hashes and computes the set of parameters that point to different constants.
 - `merge`: Uses the stable function map to optimistically create a merged function.

We can enable a global merging mode similar to the global function outliner (https://discourse.llvm.org/t/rfc-enhanced-machine-outliner-part-2-thinlto-nolto/78753/), which will perform the above steps separately.
 - `-codegen-data-generate`: During the first round of code generation, we analyze local merging instances and publish their summaries.
 - Offline using `llvm-cgdata` or at link-time, we can finalize all these merging summaries that are combined to determine parameters.
 - `-codegen-data-use`: During the second round of code generation, we optimistically create merging instances within each module, and finally, the linker folds identically created merging instances.

Depends on #112664
This is a patch for https://discourse.llvm.org/t/rfc-global-function-merging/82608.

kyulee-com committed:
[CGData] Global Merge Functions

tschuett commented on llvm/include/llvm/Transforms/IPO/GlobalMergeFunctions.h:
```diff
@@ -0,0 +1,77 @@
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef PIKA_TRANSFORMS_UTILS_GLOBALMERGEFUNCTIONS_H
```
`PIKA`

tschuett commented on llvm/include/llvm/Transforms/IPO/GlobalMergeFunctions.h:
```diff
@@ -0,0 +1,77 @@
+/// GlobalMergeFunc finds functions which only differ by constants in
+/// certain instructions, e.g. resulting from specialized functions of layout
+/// compatible types.
+/// Unlike PikaMergeFunc that directly compares IRs, this uses stable function
```
`PikaMergeFunc`

kyulee-com committed:
Address comments from tschuett

kyulee-com commented:
cc. @nocchijiang

kyulee-com commented:
cc. @eeckstein

eeckstein commented:
Nice! Will this new pass make Swift's function merging pass obsolete?

kyulee-com commented:
> Nice! Will this new pass make Swift's function merging pass obsolete?

Yes. We use this new pass with thin-LTO, which is built on top of Swift's existing function merging pass. Since the former fully covers the functionality of the latter, technically we can disable it.

ellishg commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+    bool Identical = true;
+    for (unsigned J = 1; J < StableFunctionCount; ++J) {
+      auto &SF = SFS[J];
+      assert(SF->IndexOperandHashMap->count(IndexPair));
+      auto SHash = (*SF->IndexOperandHashMap)[IndexPair];
```
`DenseMap::at()` will abort if the key doesn't exist.
```suggestion
      auto SHash = SF->IndexOperandHashMap->at(IndexPair);
```

ellishg commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+  ParamLocsVecTy ParamLocsVec;
+  for (auto &[HashSeq, Locs] : HashSeqToLocs) {
+    ParamLocsVec.push_back(std::move(Locs));
+    std::sort(
+        ParamLocsVec.begin(), ParamLocsVec.end(),
+        [&](const ParamLocs &L, const ParamLocs &R) { return L[0] < R[0]; });
```
Should we use `stable_sort()`?

```suggestion
    llvm::sort(
        ParamLocsVec,
        [](const ParamLocs &L, const ParamLocs &R) { return L[0] < R[0]; });
```

kyulee-com replied:
stable_sort is not needed as the locations should be unique.

ellishg commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+    // Compute the parameter locations based on the unique hash sequences
+    // across the candidates.
+    auto ParamLocsVec = computeParamInfo(SFS);
+    LLVM_DEBUG({
+      dbgs() << "[GlobalMergeFunc] Merging hash: " << Hash << " with Params "
+             << ParamLocsVec.size() << "\n";
+    });
```
```suggestion
    LLVM_DEBUG(
      dbgs() << "[GlobalMergeFunc] Merging hash: " << Hash << " with Params "
             << ParamLocsVec.size() << "\n"
    );
```

ellishg commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+        MergedModId = *FunctionMap->getNameForId(SF->ModuleNameId);
+      }
+
+      FuncMergeInfos.push_back({SF.get(), F, std::move(FI.IndexInstruction)});
```
```suggestion
      FuncMergeInfos.emplace_back(SF.get(), F, std::move(FI.IndexInstruction));
```

ellishg commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+    if (FuncMergeInfoSize == 0)
+      continue;
+
+    LLVM_DEBUG({
+      dbgs() << "[GlobalMergeFunc] Merging function count " << FuncMergeInfoSize
+             << " in  " << ModId << "\n";
+    });
```
```suggestion
    LLVM_DEBUG(
      dbgs() << "[GlobalMergeFunc] Merging function count " << FuncMergeInfoSize
             << " in  " << ModId << "\n"
    );
```

ellishg commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+} // namespace llvm
+
+void GlobalMergeFunc::initializeMergerMode(const Module &M) {
+  LocalFunctionMap = std::make_unique<StableFunctionMap>();
+
+  if (DisableGlobalMerging)
+    return;
```
Should we swap these lines and return early if disabled?

kyulee-com replied:
Nope. It's intentional. This disabling flag is only for the global merging part while the local merging is still enabled. I added comments.

ellishg commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+}
+
+void GlobalMergeFunc::emitFunctionMap(Module &M) {
+  LLVM_DEBUG({
+    dbgs() << "Emit function map. Size: " << LocalFunctionMap->size() << "\n";
+  });
```
```suggestion
  LLVM_DEBUG(
    dbgs() << "Emit function map. Size: " << LocalFunctionMap->size() << "\n"
  );
```

nocchijiang commented:
Shall we add new pass manager support in this PR or you have planned a separated one for it?

nocchijiang commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+  NewFunction->setDLLStorageClass(GlobalValue::DefaultStorageClass);
+
+  NewFunction->setLinkage(GlobalValue::InternalLinkage);
+  NewFunction->addFnAttr(Attribute::NoInline);
```
If `MergedFunc` is `alwaysinline`, we should remove the attribute from `NewFunction`. The verifier will complain about it if both `noinline` and `alwaysinline` are present.

kyulee-com replied:
I bailed out `alawaysinline` case from the eligible function as it doesn't make sense to merge functions that must be inlined.

nocchijiang commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+    // objc_msgSend stubs must be called, and can't have their address taken.
+    if (Callee->getName().starts_with("objc_msgSend$"))
+      return false;
+  }
```
Calls to dtrace probes should probably be filtered here as well. Linkers have special logic to handle them. Parameterizing these calls will break the linkage. See 6c641d0de686efd3e3da055aa4c743bb3450acb7 for lld, https://github.com/apple-oss-distributions/ld64/blob/47f477cb721755419018f7530038b272e9d0cdea/src/ld/OutputFile.cpp#L1597 for ld64.

kyulee-com replied:
Yeah. I now bail out `__dtrace` case similar to https://github.com/swiftlang/swift/blob/main/lib/LLVMPasses/LLVMMergeFunctions.cpp#L643-L644.

tschuett commented:
Could please add one or more rst files to https://github.com/llvm/llvm-project/tree/main/llvm/docs about cgdata.
Tools, passes, flags, and more.
For better visibility.

kyulee-com committed:
address comments from ellishg

kyulee-com committed:
Address comments from nocchijiang

nocchijiang commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+  for (auto &[Hash, SFS] : FunctionMap->getFunctionMap()) {
+    // Compute the parameter locations based on the unique hash sequences
+    // across the candidates.
+    auto ParamLocsVec = computeParamInfo(SFS);
```
This seems precomputable during `llvm-cgdata` invocation, or at least we can lazily calculate it on demand.

kyulee-com replied:
In fact, I've moved several pieces of logic from the offline phase (llvm-cgdata) into the merging step for simplicity. Admittedly, this approach is less efficient. I plan to reorganize some of the more obvious parts in the next update to this PR.

nocchijiang commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+  DenseSet<Function *> MergedFunctions;
+
+  auto ModId = M.getModuleIdentifier();
+  for (auto &[Hash, SFS] : FunctionMap->getFunctionMap()) {
```
Can we compute hashes for local functions then lookup the global map by hashes? A full iteration seems to be too much waste especially for non-LTO builds.

kyulee-com replied:
That's a good point. For the similar reason above, I did this way for simplicity. We may restructure the code a bit in this case. Will leave it out for this PR.

kyulee-com replied:
@nocchijiang I attempted to refactor the code using #115750 rather than modifying it directly in this PR. The potential catch is that the new method might be overly optimistic in certain scenarios, which could lead to an increase in size. However, it should not affect the correctness. And also I didn't fully test actual large cases. Let me know your thought.

nocchijiang replied:
I just took a look at the draft, overall lgtm. But I am not 100% confident that I got the point of the potential size increase, correct me if I am wrong.

nocchijiang commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+      // Get the function from the stable name.
+      auto I = StableNameToFuncMap.find(
+          *FunctionMap->getNameForId(SF->FunctionNameId));
+      if (I == StableNameToFuncMap.end())
```
The merging requires an almost exact name match in the global record to proceed? Does this mean a new function that is theoretically mergable will never be merged until a newer CGData is generated & fed into next compilation?

kyulee-com replied:
Yes, it appears that the current approach doesn't merge new functions with different names, which seem worth improving.

nocchijiang commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+        // Check if the matched functions fall into the same (first) module.
+        // This module check is not strictly necessary as the functions can move
+        // around. We just want to avoid merging functions from different
+        // modules than the first one in the functon map, as they may not end up
```
```suggestion
        // modules than the first one in the function map, as they may not end up
```

nocchijiang commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+        // This module check is not strictly necessary as the functions can move
+        // around. We just want to avoid merging functions from different
+        // modules than the first one in the functon map, as they may not end up
+        // with not being ICFed by the linker.
```
IIUC:

```suggestion
        // with being ICFed by the linker.
```

nocchijiang reviewed (commented):
This batch of comments acts more like some questions for me to better understand the design decisions.

kyulee-com committed:
Fix format

kyulee-com commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+  default:
+    return false;
+  }
+}
```
Nice catch! These are duplicated, which I forgot cleaning them up.

kyulee-com commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+  // Add new arguments defined by Params.
+  for (auto *Param : Params) {
+    assert(ParamIdx < ToFuncTy->getNumParams());
+    // FIXME: do not support signing
```
This is about support for a target that enables pointer authentication, which appears to be supported by the Swift function merger. In the current implementation, I've opted to bypass this case for now. Nonetheless, I've removed the comment since the current implementation doesn't support it anyway.

kyulee-com commented:
> Shall we add new pass manager support in this PR or you have planned a separated one for it?

I refactor the pass to support both legacy and new pass manager. Also directly wire it into a codegen pass.

kyulee-com commented:
> Could please add one or more rst files to https://github.com/llvm/llvm-project/tree/main/llvm/docs about cgdata. Tools, passes, flags, and more. For better visibility.

We already have a few docs available:
 - Clang flags about `-fcodegen-data-generate`and `-fcodegen-data-use`: https://github.com/llvm/llvm-project/blob/main/clang/docs/UsersManual.rst
 - `llvm-cgdata` tool: https://github.com/llvm/llvm-project/blob/main/llvm/docs/CommandGuide/llvm-cgdata.rst
 
However, these are somewhat fragmented. Once all flags and linker support are finalized, I plan to create a separate rst document for cgdata, including more examples and usage guidelines.

kyulee-com commented on llvm/lib/Transforms/IPO/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,744 @@
+  // Build a map from stable function name to function.
+  StringMap<Function *> StableNameToFuncMap;
+  for (auto &F : M)
+    StableNameToFuncMap[get_stable_name(F.getName())] = &F;
```
I think function names are unique within a module. We mostly introduce suffixes to identify them from different modules.

thevinster commented on llvm/lib/CodeGen/TargetPassConfig.cpp:
```diff
@@ -887,6 +891,9 @@ void TargetPassConfig::addIRPasses() {
     addPass(createSelectOptimizePass());
+
+  if (EnableGlobalMergeFunc)
+    addPass(createGlobalMergeFuncPass());
```
Were there any principled judgment for placing placing the pass here?

thevinster commented on llvm/lib/CodeGen/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,732 @@
+using namespace llvm::support;
+
+static cl::opt<bool>
+    DisableGlobalMerging("disable-global-merging", cl::Hidden,
```
Would the name here be confusing? i.e. Readers might be assume the functionality is no-op when passing both `-enable-global-merge-func` and `disable-global-merging`. Perhaps something like `-disable-codegen-data`?

kyulee-com committed:
Change related to 2nd comments from nocchijiang

- Move isProfitable to offline (llvm-cgdata)
- skip-trim flag to skip trimming codegen data.
  - The default behavior is to trim unnecessary content from the codegen
    data.
  - With this flag enabled, we skip trimming, which is often useful to
    create a synthetic test.

kyulee-com committed:
Address comments from thevinster

mikaelholmen commented:
Hi,

I've no idea if this is expected but I noticed that if I run
```opt -passes=global-merge-func gmf.ll -o /dev/null```
on gmf.ll being
```
@0 = global { { ptr, i32, i32 } } { { ptr, i32, i32 } { ptr null, i32 19, i32 5 } }
@1 = global { { ptr, i32, i32 } } { { ptr, i32, i32 } { ptr null, i32 22, i32 5 } }
```
it crashes like
```
opt: ../lib/Analysis/ModuleSummaryAnalysis.cpp:1072: ModuleSummaryIndex llvm::buildModuleSummaryIndex(const Module &, std::function<BlockFrequencyInfo *(const Function &)>, ProfileSummaryInfo *, std::function<const StackSafetyInfo *(const Function &)>): Assertion `GlobalList.second.SummaryList.size() == 1 && "Expected module's index to have one summary per GUID"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: build-all/bin/opt -passes=global-merge-func gmf.ll -o /dev/null
1.	Running pass "global-merge-func" on module "gmf.ll"
 #0 0x000055e51084dd78 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/opt+0x4427d78)
 #1 0x000055e51084b83e llvm::sys::RunSignalHandlers() (build-all/bin/opt+0x442583e)
 #2 0x000055e51084e5ad SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f1cdb046cf0 __restore_rt (/lib64/libpthread.so.0+0x12cf0)
 #4 0x00007f1cd8bffacf raise (/lib64/libc.so.6+0x4eacf)
 #5 0x00007f1cd8bd2ea5 abort (/lib64/libc.so.6+0x21ea5)
 #6 0x00007f1cd8bd2d79 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d79)
 #7 0x00007f1cd8bf8426 (/lib64/libc.so.6+0x47426)
 #8 0x000055e511442fe1 llvm::buildModuleSummaryIndex(llvm::Module const&, std::function<llvm::BlockFrequencyInfo* (llvm::Function const&)>, llvm::ProfileSummaryInfo*, std::function<llvm::StackSafetyInfo const* (llvm::Function const&)>) (build-all/bin/opt+0x501cfe1)
 #9 0x000055e511447e72 llvm::ModuleSummaryIndexAnalysis::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x5021e72)
#10 0x000055e5123dce42 llvm::detail::AnalysisPassModel<llvm::Module, llvm::ModuleSummaryIndexAnalysis, llvm::AnalysisManager<llvm::Module>::Invalidator>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) PassBuilder.cpp:0:0
#11 0x000055e510a5849a llvm::AnalysisManager<llvm::Module>::getResultImpl(llvm::AnalysisKey*, llvm::Module&) (build-all/bin/opt+0x463249a)
#12 0x000055e511b31fbf llvm::GlobalMergeFuncPass::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x570bfbf)
#13 0x000055e5123ef72d llvm::detail::PassModel<llvm::Module, llvm::GlobalMergeFuncPass, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) PassBuilder.cpp:0:0
#14 0x000055e510a55477 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x462f477)
#15 0x000055e511b9f6f3 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (build-all/bin/opt+0x57796f3)
#16 0x000055e51081514b optMain (build-all/bin/opt+0x43ef14b)
#17 0x00007f1cd8bebd85 __libc_start_main (/lib64/libc.so.6+0x3ad85)
#18 0x000055e51080ef6e _start (build-all/bin/opt+0x43e8f6e)
Abort (core dumped)
```

kyulee-com commented:
> Hi,
> 
> I've no idea if this is expected but I noticed that if I run `opt -passes=global-merge-func gmf.ll -o /dev/null` on gmf.ll being
> 
> ```
> @0 = global { { ptr, i32, i32 } } { { ptr, i32, i32 } { ptr null, i32 19, i32 5 } }
> @1 = global { { ptr, i32, i32 } } { { ptr, i32, i32 } { ptr null, i32 22, i32 5 } }
> ```
> 
> it crashes like

Thank you for testing this! I can't reproduce the issue on my end, but I suspect it might be due to the module summary being run twice, even though it's optional.
Could you please try out https://github.com/llvm/llvm-project/pull/116241 again?

mikaelholmen commented:
> > Hi,
> > I've no idea if this is expected but I noticed that if I run `opt -passes=global-merge-func gmf.ll -o /dev/null` on gmf.ll being
> > ```
> > @0 = global { { ptr, i32, i32 } } { { ptr, i32, i32 } { ptr null, i32 19, i32 5 } }
> > @1 = global { { ptr, i32, i32 } } { { ptr, i32, i32 } { ptr null, i32 22, i32 5 } }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > it crashes like
> 
> Thank you for testing this! I can't reproduce the issue on my end, but I suspect it might be due to the module summary being run twice, even though it's optional. Could you please try out #116241 again?

Weird if you can't reproduce it. I did again at 878b03e0b9669. Asserts must be on.
Anyway I tried #116241 and with that it doesn't crash for me.

kyulee-com commented:
> 
> Weird if you can't reproduce it. I did again at [878b03e](https://github.com/llvm/llvm-project/commit/878b03e0b96698ced5fb6a70dc80df05ef884f8c). Asserts must be on. Anyway I tried #116241 and with that it doesn't crash for me.

Yeah. I forgot enabling asserts. Now I can see an assertion. I think #116241 should land it for this fix. Can you review it?

nocchijiang commented on llvm/lib/CGData/StableFunctionMap.cpp:
```diff
@@ -117,7 +149,38 @@ static void removeIdenticalIndexPair(
+
+  unsigned Benefit = InstCount * (StableFunctionCount - 1);
+  unsigned Cost =
+      (GlobalMergingParamOverhead * ParamCount + GlobalMergingCallOverhead) *
```
Also FYI the implementation that comes with this PR is not passing all the existing Swift repo tests despite lowering `GlobalMergingParamOverhead` to `0` to force merges to happen.

kyulee-com replied:
>  it overestimated the cost for some parameters of small scalar type.

It seems we need to incorporate some type information to accurately reflect the precise cost, in theory. Since the profit model is currently computed offline, I've refined the parameter count for greater precision, as mentioned above

> FYI the implementation that comes with this PR is not passing all the existing Swift repo tests despite lowering GlobalMergingParamOverhead to 0 to force merges to happen.

Despite some differences in the underlying assumptions, I'm curious about what the existing Swift merge can accomplish that this new pass cannot.

nocchijiang replied:
> I'm curious about what the existing Swift merge can accomplish that this new pass cannot.

These are the behavioral differences I have found by running `merge_func*.ll` from Swift repo, some of which I believe we should consider implementing as well:

|Test|Differences|Remark|
|----|-----------|--------|
|`merge_func_preserves_vfe.ll`|`merge_candidate_c` merged or not|The new pass seems unaware of metadata differences.|
|`merge_func_ptrauth.ll`|presence of ptrauth info|I believe the new pass does not implement ptrauth for simplicity.|
|`merge_func_return_type_cast.ll`|`return_0` merged with `return_null` or not|The 2 functions have different stable hash values.|
|`merge_func.ll`|`func*_merged_with*` merging strategy|Apple implementation tends to avoid too many parameters.|
|`merge_func.ll`|`caller*_*` merging strategy|Apple implementation does not parameterize the calls on the call chain.|
|`merge_func.ll`|`first` merged with `second` or not|Apple implementation is aware of the order of incoming blocks to a phi.|
|`merge_func.ll`|`not_really_recursive` eliminated or not|Apple implementation replaces the call to the original function with the merged one.|

nocchijiang commented on llvm/lib/CodeGen/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,672 @@
+
+      // Convert the operand map to a vector for a serialization-friendly
+      // format.
+      IndexOperandHashVecType IndexOperandHashes;
```
Should we bail out for functions without parameterizable constants here?

kyulee-com replied:
I've found not skipping no parameter case is beneficial. Anyhow, I now introduced `-global-merging-skip-no-params` to control this behavior.

nocchijiang reviewed (commented):
Sorry for some late comments, since I didn't have free time actually debugging around the changes.

kyulee-com commented on llvm/lib/CodeGen/GlobalMergeFunctions.cpp:
```diff
@@ -0,0 +1,672 @@
+  ParamLocsVecTy ParamLocsVec;
+  for (auto &[HashSeq, Locs] : HashSeqToLocs) {
+    ParamLocsVec.push_back(std::move(Locs));
+    llvm::sort(ParamLocsVec, [&](const ParamLocs &L, const ParamLocs &R) {
```
Nice catch! Apparently I mis-placed this code inside the loop while I was refactoring this code.

kyulee-com commented on llvm/lib/CGData/StableFunctionMap.cpp:
```diff
@@ -117,7 +149,38 @@ static void removeIdenticalIndexPair(
+  if (StableFunctionCount < GlobalMergingMinMerges)
+    return false;
+
+  unsigned InstCount = SFS[0]->InstCount;
```
I added `-global-merging-inst-overhead` to tune this parameter.

kyulee-com commented on llvm/lib/CGData/StableFunctionMap.cpp:
```diff
@@ -117,7 +149,38 @@ static void removeIdenticalIndexPair(
+  if (InstCount < GlobalMergingMinInstrs)
+    return false;
+
+  unsigned ParamCount = SFS[0]->IndexOperandHashMap->size();
```
The number of unique values can be different depending on a stable function.
Anyhow, I consider this count as a parameter count, and add them up to compute the total cost.

kyulee-com commented:
> Sorry for some late comments, since I didn't have free time actually debugging around the changes.

Thank you for testing and reporting these issues!
I now incorporated them into #116548.
